src/librustc_yk_sections/emit_tir.rs

```diff
@@ -262,6 +306,108 @@ fn insert_phi(block: &mut ykpack::BasicBlock, var: TirLocal, arity: usize) {
     block.stmts.insert(0, ykpack::Statement::Assign(lhs, rhs));
 }
 
+/// A statement location.
+#[derive(Clone, Debug)]
+struct StmtLoc {
+    /// The block containing the statement.
+    bb: TirBasicBlockIndex,
+    /// The statement index.
+    si: usize,
+}
+
+/// SSA variable renaming. Algorithm reference:
+/// Bottom of p408 of 'Modern Compiler Implementation in Java (2nd ed.)' by Andrew Appel.
+struct RenameCx {
+    /// A counter used to give new TIR variables a unique identifier.
+    /// The original algorithm used one counter per original variable. This would mean storing each
+    /// SSA variable as a (name, version) pair. For added efficiency, we use a single counter and
+    /// represent our variables as plain old integers.
+    count: TirLocal,
+    /// Each variable has a stack of definitions.
+    stack: Vec<Vec<TirLocal>>,
+}
+
+impl RenameCx {
+    /// Make a new renaming context. To prevent variable naming clashes, the `next_fresh_var`
+    /// argument should be one more than the last variable the previous step of the conversion
+    /// created.
+    fn new(next_fresh_var: u32) -> Self {
+        // We start with space for the variables we know about so far. The vectors will grow as new
+        // SSA variables are instantiated.
+        let next_fresh_var_usize = usize::try_from(next_fresh_var).unwrap();
+        let mut stack = Vec::with_capacity(next_fresh_var_usize);
+        stack.resize(next_fresh_var_usize, Vec::new());
+        Self {
+            count: next_fresh_var,
+            stack,
+        }
+    }
+
+    /// Create a new SSA variable.
+    fn fresh_var(&mut self) -> TirLocal {
+        let ret = self.count;
+        self.count = self.count.checked_add(1).unwrap();
+        ret
+    }
+
+    /// Entry point for variable renaming.
+    fn rename_all(mut self, doms: &Dominators<BasicBlock>, mir: &Mir,
+        blks: &mut Vec<ykpack::BasicBlock>)
+    {
+        // We start renaming in the entry block and it ripples down the dominator tree.
+        self.rename(doms, mir, blks, 0);
+    }
+
+    // FIXME rename variables in terminators.
+    fn rename(&mut self, doms: &Dominators<BasicBlock>, mir: &Mir,
+        blks: &mut Vec<ykpack::BasicBlock>, n: TirBasicBlockIndex)
+    {
+        let n_usize = n as usize;
+        // We have to remember the variables whose stacks we must pop from when we come back from
+        // recursion. These must be the variables *before* they were renamed.
+        let mut pop_later = Vec::new();
```

Comment:
Can a variable appear more than once in here? Does it happen often that variables end up in here?

Reply:
Every variable assignment in the block we are processing will push to this vector. The same variable may appear more than once if a block assigns to the same variable many times.

Reply:
So this should probably be a set to avoid (I assume pointlessly) poping the same thing over and over again?

Reply:
If a variable is assigned to X times, then we need to pop X times.

The the algorithm recursively walks the dominator tree depth-first. When it sees an assignment to a variable, it makes a new SSA variable, rewrites the old variable. and pushes the new variable onto the stack of the corresponding old variable. When we are bubbling up from recursion, we have to revert to the previous SSA versions of the variables. So if a block assigned to a variable `z` twice, we'd need to revert to `z` as of two assignments ago, thus popping twice.

Hard to explain...

Reply:
Got it.